### PR TITLE
Fixes issue-3199: Parsing atomtypes section of itp files

### LINKED ISSUE #3199: ITPParser gets confused with masses
## Expected behavior ##

MDAnalysis is capable of reading itp-style topology files in the same way gromacs does. Getting the masses of all atoms right.

## Actual behavior ##

If some masses are not defined inside the `[ atoms ]` section, the parser disregards all of them and tries guessing.

---

MDAnalysis looks for masses in the `[ atoms ]` section of the `[ moleculetype ]` section. It disregards the `[ atomtypes ]` section. That a limitation, but not a very big problem.

If one happens to have just one atom that has its mass implied from the `[ atomtypes ]` section instead of defining it directly, the parser throws out all masses and tries to guess them all, even those defined in the right spot. 

At the very least there should be a warning about discarding the masses and guessing them. When I first ran into this, I couldn't make sense of it that MDAnalysis tried to guess masses it should know.
But I think it would be elegant to only guess the masses that are not available. 
And the best case would be to infer them from the `[ atomtypes ]` section, completely mimicking gromacs.

## Code to reproduce the behavior ##

I attached a stub topology I used to test this.

``` python
$ python3 -c 'import MDAnalysis as mda; u = mda.Universe("topol.top", topology_format="itp"); print(u.atoms.atoms.masses)'#
/home/henrik/mdanalysis/package/MDAnalysis/topology/guessers.py:146: UserWarning: Failed to guess the mass for the following atom types: AY1
  warnings.warn("Failed to guess the mass for the following atom types: {}".format(atom_type))
/home/henrik/mdanalysis/package/MDAnalysis/topology/guessers.py:146: UserWarning: Failed to guess the mass for the following atom types: AY2
  warnings.warn("Failed to guess the mass for the following atom types: {}".format(atom_type))
/home/henrik/mdanalysis/package/MDAnalysis/topology/guessers.py:146: UserWarning: Failed to guess the mass for the following atom types: HOY
  warnings.warn("Failed to guess the mass for the following atom types: {}".format(atom_type))
/home/henrik/mdanalysis/package/MDAnalysis/topology/guessers.py:146: UserWarning: Failed to guess the mass for the following atom types: K+
  warnings.warn("Failed to guess the mass for the following atom types: {}".format(atom_type))
/home/henrik/mdanalysis/package/MDAnalysis/topology/guessers.py:146: UserWarning: Failed to guess the mass for the following atom types: OY1
  warnings.warn("Failed to guess the mass for the following atom types: {}".format(atom_type))
/home/henrik/mdanalysis/package/MDAnalysis/topology/guessers.py:146: UserWarning: Failed to guess the mass for the following atom types: OY2
  warnings.warn("Failed to guess the mass for the following atom types: {}".format(atom_type))
/home/henrik/mdanalysis/package/MDAnalysis/topology/guessers.py:146: UserWarning: Failed to guess the mass for the following atom types: OY3
  warnings.warn("Failed to guess the mass for the following atom types: {}".format(atom_type))
/home/henrik/mdanalysis/package/MDAnalysis/topology/guessers.py:146: UserWarning: Failed to guess the mass for the following atom types: OY4
  warnings.warn("Failed to guess the mass for the following atom types: {}".format(atom_type))
/home/henrik/mdanalysis/package/MDAnalysis/topology/guessers.py:146: UserWarning: Failed to guess the mass for the following atom types: OY5
  warnings.warn("Failed to guess the mass for the following atom types: {}".format(atom_type))
/home/henrik/mdanalysis/package/MDAnalysis/topology/guessers.py:146: UserWarning: Failed to guess the mass for the following atom types: OY6
  warnings.warn("Failed to guess the mass for the following atom types: {}".format(atom_type))
/home/henrik/mdanalysis/package/MDAnalysis/topology/guessers.py:146: UserWarning: Failed to guess the mass for the following atom types: OY7
  warnings.warn("Failed to guess the mass for the following atom types: {}".format(atom_type))
/home/henrik/mdanalysis/package/MDAnalysis/topology/guessers.py:146: UserWarning: Failed to guess the mass for the following atom types: OY8
  warnings.warn("Failed to guess the mass for the following atom types: {}".format(atom_type))
/home/henrik/mdanalysis/package/MDAnalysis/topology/guessers.py:146: UserWarning: Failed to guess the mass for the following atom types: OY9
  warnings.warn("Failed to guess the mass for the following atom types: {}".format(atom_type))
/home/henrik/mdanalysis/package/MDAnalysis/topology/guessers.py:146: UserWarning: Failed to guess the mass for the following atom types: SY1
  warnings.warn("Failed to guess the mass for the following atom types: {}".format(atom_type))
/home/henrik/mdanalysis/package/MDAnalysis/topology/guessers.py:146: UserWarning: Failed to guess the mass for the following atom types: SY2
  warnings.warn("Failed to guess the mass for the following atom types: {}".format(atom_type))
/home/henrik/mdanalysis/package/MDAnalysis/core/universe.py:157: UserWarning: No coordinate reader found for topol.top. Skipping this file.
  warnings.warn('No coordinate reader found for {}. Skipping '
[0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.
 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.
 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.
 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]

....

```

## Current version of MDAnalysis ##

```
$ python3 -c "import MDAnalysis as mda; print(mda.__version__)"
2.0.0-dev0
$ python3 -V
Python 3.8.5
$ lsb_release -a
No LSB modules are available.
Distributor ID:	Ubuntu
Description:	Ubuntu 20.04.2 LTS
Release:	20.04
Codename:	focal
```

## topology ##

```
[ defaults ]
; nbfunc        comb-rule       gen-pairs       fudgeLJ fudgeQQ
  1             2               no              1.0     1.0

; ions       
[ atomtypes ]
;name   at.num  mass      charge  ptype   sigma  epsilon
K       19      39.098    1       A       0.283  0.90
NA      11      22.990    1       A       0.231  0.45
CL      17      35.450    -1      A       0.430  0.42
BR      35      10.810    -1      A       0.443  0.75

; Mica
[ atomtypes ]
;name   at.num  mass        charge          ptype       sigma           epsilon
SY1         14      28.08600    0.000000        A       0.356359487256  0.209200
AYT2        13      26.98153    0.000000        A       0.374177461619  0.209200
OY1          8      15.99940    0.000000        A       0.311814551349  0.104600
OY3          8      15.99940    0.000000        A       0.311814551349  0.104600
OY5          8      15.99940    0.000000        A       0.311814551349  0.104600
AY2         13      26.98153    0.000000        A       0.374177461619  0.209200
OY7          8      15.99940    0.000000        A       0.311814551349  0.104600
OY2          8      15.99940    0.000000        A       0.311814551349  0.104600
HOY          1      1.008000    0.000000        A       0.096662510918  0.062760
AY1         13      26.98153    0.000000        A       0.374177461619  0.209200
OY8          8      15.99940    0.000000        A       0.311814551349  0.104600
K+          19      39.09830    0.000000        A       0.338541512893  0.836800
OY6          8      15.99940    0.000000        A       0.311814551349  0.104600
AYT1        13      26.98153    0.000000        A       0.374177461619  0.209200
SY2         14      28.08600    0.000000        A       0.356359487256  0.209200
OY9          8      15.99940    0.000000        A       0.311814551349  0.104600
OY4          8      15.99940    0.000000        A       0.311814551349  0.104600

[ moleculetype ]
; molname    nrexcl
SURF    3

[ atoms ]
;   nr       type  resnr residue  atom   cgnr     charge       mass
     1        AY1      1   SURF     Al      1      1.449     26.982   ;
     2        SY1      1   SURF     Si      2      1.100     28.086   ;
     3        SY2      1   SURF     Si      3      1.100     28.086   ;
     4        HOY      1   SURF      H      4      0.200      1.008   ;
     5        OY1      1   SURF      O      5     -0.550     15.999   ;
     6        OY2      1   SURF      O      6     -0.550     15.999   ;
     7        OY3      1   SURF      O      7     -0.550     15.999   ;
     8        OY7      1   SURF      O      8     -0.758     15.999   ;
     9        OY8      1   SURF      O      9     -0.758     15.999   ;
    10        OY9      1   SURF      O     10     -0.683     15.999   ;
    11        AY1      1   SURF     Al     11      1.449     26.982   ;
    12        SY1      1   SURF     Si     12      1.100     28.086   ;
    13        SY2      1   SURF     Si     13      1.100     28.086   ;
    14        HOY      1   SURF      H     14      0.200      1.008   ;
    15        OY1      1   SURF      O     15     -0.783     15.999   ;
    16        OY2      1   SURF      O     16     -0.783     15.999   ;
    17        OY3      1   SURF      O     17     -0.550     15.999   ;
    18        OY7      1   SURF      O     18     -0.758     15.999   ;
    19        OY8      1   SURF      O     19     -0.758     15.999   ;
    20        OY9      1   SURF      O     20     -0.683     15.999   ;
    21        AY2      1   SURF     Al     21      1.448     26.982   ;
    22        AY2      1   SURF     Al     22      1.448     26.982   ;
    23        SY1      1   SURF     Si     23      1.100     28.086   ;
    24        SY2      1   SURF     Si     24      1.100     28.086   ;
    25        HOY      1   SURF      H     25      0.200      1.008   ;
    26        OY1      1   SURF      O     26     -0.783     15.999   ;
    27        OY2      1   SURF      O     27     -0.550     15.999   ;
    28        OY3      1   SURF      O     28     -0.550     15.999   ;
    29        OY4      1   SURF      O     29     -0.758     15.999   ;
    30        OY5      1   SURF      O     30     -0.758     15.999   ;
    31        OY6      1   SURF      O     31     -0.683     15.999   ;
    32        AY1      1   SURF     Al     32      1.449     26.982   ;

[ moleculetype ]
CION                1
[ atoms ]
1       K+          1       CION        K+       1      1 

[ system ]
Mica surface with specific water-salt mixture

[ molecules ]
SURF     1 ; periodic slab
CION	60
```

### PR DESCRIPTION
**Pull request by hejamu**
The info in atomtypes is written to a dict. It is then used for the masses not found in the [ atoms ] section. Masses that are still not found (missing include file could be a reason), the masses are guessed based on the guessed atomtype.

Fixes #3199

Changes made in this Pull Request:
 - The info in atomtypes is written to a dict. 
 - It is then used for the masses not found in the [ atoms ] section. 
 - Masses that are still not found (missing include file could be a reason), the masses are guessed based on the guessed atomtype.

PR Checklist
------------
 - [x] Tests?
 - [x] Docs?
 - [x] CHANGELOG updated?
 - [x] Issue raised/referenced?

### CODE REVIEW

**Comment by hejamu** [comment]
Oh it is the `guessed` attribute it is worried about. That should be an easy fix.

**Comment by hejamu** [comment]
Just leaving a thought here, not at the desk right now:

The same thing would apply to charges, wouldn’t it?

**Comment by hejamu** [comment]
Let us imagine a topology like so

```
[ defaults ]
; nbfunc        comb-rule       gen-pairs       fudgeLJ fudgeQQ
  1             2               no              1.0     1.0

#include "no_longer_avail.file"

[ atomtypes ]
;name  at.num  mass        charge          ptype       sigma           epsilon
A         14         10.086     2.000        A       0.356359487256  0.209200
B         13         20.989    -3.000        A       0.374177461619  0.209200

[ moleculetype ]
; molname    nrexcl
Test    1

[ atoms ]
;nr      type  resnr residue  atom   cgnr    charge    mass
1        A     1     SURF      A      1      4.000     8.000;
2        B     1     SURF      B      2      1.100     ;
3        B     1     SURF      B      3      ;
4        H     1     SURF      H      4      ;

[ system ]
Testing

[ molecules ]
TEST     1
```

With the atomtype `H` being defined in some forcefield file.

Gromacs will happily read that in with the following parameters

```
Nr Type Charge         Mass
1   A   4.000         8.000
2   B   1.100         20.980
3   B   -3.00         20.980
4   H   (Whatever is listed in the no_longer_avail.file)
```

MDAnalysis should basically mimic this behavior, like the code in the PR does but also for charges.

One thing that doesn't make sense any longer then is to let the parser fail in case a include file is missing. (relates to  PR #3108 )

If all include files are present, all values are defined and there is no need to guess any, like it is done now. 

If however, a include file is missing, mostly the standard force fields I imagine (because custom FF files are mostly send along if one sends a system to anyone), the guesser needs to look the values up (and because they are mostly standard force fields that should mostly work).

___

So I propose the following, correct me if this is a bad idea:

- Masses and charges should be parsed in the way shown for masses in this PR.
- The Parser should not fail on missing include files, rather raise a warning and proceed, using the guesser to fill the missing values.
- This would mimic the behavior of gromacs exactly, while still working if some force field info is not present on the system.

This is still a bug rather than a feature request, because right now MDAnalysis does not take the correct masses and charges from the topology file in some cases, either dismissing them or guessing them (impact might be minuscule there, but still)

**Comment by IAlibay** [comment]
> The Parser should not fail on missing include files, rather raise a warning and proceed, using the guesser to fill the missing values.

Apologies, I've not been keeping up too much with this PR. May I ask for a clarification of to what extent we are guessing here?
Generally mass guessing has been acceptable in MDA, but we have been really trying to avoid guessing where possible (since guaranteeing accuracy has always been a problem). I personally would say that charge guessing might be out of scope of what we can do.

I'm willing to be open minded here though, how would you propose charge guessing in the case of a missing itp file? (i.e. where/how are you matching with the standard ff, how do you avoid not storing the entirety of the gmx force field library?)

**Comment by hejamu** [comment]
Sorry, I didn't look it up and was under the impression MDA also has a charge guessing routine. Ignore the charge guessing then, this is really out of the scope.

But do you/the rest of the devs agree with the proposal, i.e. leveraging the atomtypes section for info (I really think this should be done) including for charges and not failing outright when a itp file is missing? It would be:

- Check for include files, warn if one is missing
- Check for charges and masses in [ atoms ] section
- If some are missing, check [ atomtypes ] and fill the missing info
- If charges are still missing, fail
- If masses are still missing, guess

as the workflow then.

**Comment by IAlibay** [comment]
It sounds more than reasonable to me, although I'll defer to @lilyminium's opinion here as she is more acquainted with this reader.
(note we don't necessarily have to fail if there's no charges, we could just drop them on read if we have an incomplete set)

**Comment by hejamu** [comment]
Added parsing for charges not found in the [ atoms ] section and a bit of cleanup. Also added a test for all versions of a [ atoms ] entry.

File: package/MDAnalysis/topology/ITPParser.py
Comment (lilyminium) [inline package/MDAnalysis/topology/ITPParser.py:498]:
```suggestion
        self.atomtypes = {}
```

Typo?

File: package/MDAnalysis/topology/ITPParser.py
Comment (lilyminium) [inline package/MDAnalysis/topology/ITPParser.py:542]: I am super hesitant about reading charges from atomtypes, because partial charges are so dependent on the local residue environment. Maybe we could make this a keyword argument but if there exist both an `atomtypes.atp` and an `[ atomtypes ]`, the charges of the amino acids should be taking preference. In the case where both the files exist, I think my preference would actually be

aminoacids.rtp > no charges > atomtypes

If your use case has these charges being actually accurate, I could be persuaded.

File: package/MDAnalysis/topology/ITPParser.py
Comment (lilyminium) [inline package/MDAnalysis/topology/ITPParser.py:548]: If this is too long, you could split it over two lines.

Comment (lilyminium) [reply package/MDAnalysis/topology/ITPParser.py:548]: Possibly you could even do the below to save time.
```python
try:
	m = self.atomtypes[x]["mass"]
except KeyError:
	m = guessers.guess_mass(....)
```

Not sure if a singular guess_mass function exists

File: testsuite/MDAnalysisTests/data/atomtypes.itp
Comment (lilyminium) [inline testsuite/MDAnalysisTests/data/atomtypes.itp:9]: What does the BB mean here?

Ah, I see it's a bonded type -- could you update the heading accordingly?

File: package/MDAnalysis/topology/ITPParser.py
Comment (lilyminium) [inline package/MDAnalysis/topology/ITPParser.py:620]: Could you link me to where these different specs are? The only one I see in the manual and in my own files is the one that follows the heading `;name  at.num   mass      charge  ptype       c6           c12`. Given that the lynchpin of this parsing is the particle type, is it always guaranteed to be there?

**Review by lilyminium — CHANGES_REQUESTED** [review]
Thanks @hejamu, this is a massive improvement to what we have now. I do have a few questions and concerns about the changes (elaborated in the comments below).

Regarding the proposed rules:

> - Check for include files, warn if one is missing
> - Check for charges and masses in [ atoms ] section
> - If some are missing, check [ atomtypes ] and fill the missing info
> - If charges are still missing, fail
> - If masses are still missing, guess

See my concerns about applying charges. I agree with @IAlibay that instead of failing if charges are missing, we could just not have that TopologyAttr -- charges are fairly rare in the formats we support anyway.

I will note that in my GROMOS 54A7 ``ffnonbonded.itp`` all the masses are set to 0, which for me would probably result in somewhat worse guesses than just guessing from the atom names. This is a little out of scope for the PR, but have you considered supporting `atomtypes.atp` files as well?

File: package/MDAnalysis/topology/ITPParser.py
Comment (hejamu) [reply package/MDAnalysis/topology/ITPParser.py:620]: I took them directly from the preprocessing code in gromacs, I guess that makes them as correct as possible.

https://github.com/gromacs/gromacs/blob/6dbf35dddb95932e6e98d8e923f4ba87bf257b8e/src/gromacs/gmxpreprocess/toppush.cpp#L324-L352

File: testsuite/MDAnalysisTests/data/atomtypes.itp
Comment (hejamu) [reply testsuite/MDAnalysisTests/data/atomtypes.itp:9]: Sorry, forgot that while changing the test.

File: package/MDAnalysis/topology/ITPParser.py
Comment (hejamu) [reply package/MDAnalysis/topology/ITPParser.py:498]: I have no idea what happened here, refactor stuff I guess... should be atomtypes

File: package/MDAnalysis/topology/ITPParser.py
Comment (lilyminium) [reply package/MDAnalysis/topology/ITPParser.py:620]: Yes, yes it does -- did not think to look there!

File: package/MDAnalysis/topology/ITPParser.py
Comment (hejamu) [reply package/MDAnalysis/topology/ITPParser.py:542]: I don't think this is an issue here, because the .atp/.rtp files are only relevant for topology building with pdb2gmx, but not when actually grompping and mdrunning. (I'm 99 % sure about that, but I stand to be corrected)

So at the point one has a topol.top file to work with, all charges have already been written to the [ atoms ] sections correctly.

**Comment by hejamu** [comment]
Thanks for the elaborate review, @lilyminium .

I have some problems understanding what you try to say about the forcefield files: 

If masses are set to zero for example in the ffnonbonded.itp and you use an atom from that definition without explicitly setting the mass in the [ atoms ] section, the mass of that atom will be zero. 

But that would basically be the expected behaviour in gromacs. Either you wanted zero mass for the atom or you set that incorrectly. But (in my opinion) MDAnalysis should not try to correct that user error and instead read the system in the same way the simulation software treated it.

I hope I'm not totally missing the point here.

**Comment by lilyminium** [comment]
@hejamu I do see your point and agree with it. In that case if you could fix any reasonable pep8 issues noted by the bot (https://github.com/MDAnalysis/mdanalysis/pull/3204#issuecomment-812837181) I'm happy with it. I'll just tag in @jbarnoud as well as the resident GROMACS expert, should have done that earlier.

**Comment by hejamu** [comment]
Sorry for blasting the `datafiles.py`, but my IDE was so full of PEP8 Errors that I couldn't make out my own changes.

**Comment by IAlibay** [comment]
> Sorry for blasting the `datafiles.py`, but my IDE was so full of PEP8 Errors that I couldn't make out my own changes.

@hejamu I appreciate the situation, but I would ask you to avoid doing this (especially given I've just asked the same thing of @Luthaf ).

Whilst we do encourage folks to do PEP8 changes as they alter things in the code base, large scale (i.e. whole files) changes (be they auto-formatter or not) are problematic for two reasons:

1. It muddles up the git history. Granted blame allows you to go beyond the most recent commit, but it makes it harder for us to dig into the history to see why a change was made. For something like datafiles, this is particularly important because we need to be able to track down when files were added/changes (for example for the 2021 TPR change, there was a query about the provenance of the original TPR file which took one of the coredevs some time to work out).
2. It makes it hard to review. Essentially at this point it's passing the buck of making it hard to work out changes from you to us, and we only have the git diff to work out what you're doing.

**Comment by hejamu** [comment]
Sorry, didn't think about that, but you are totally right. Reverted and made sure the new lines are compliant.

**Review by lilyminium — APPROVED** [review]

**Comment by lilyminium** [comment]
Thank you @hejamu! I'll just leave it for a day in case @jbarnoud has any comments, but otherwise LGTM.

**Pull request closed without merging**
